applications/nrf_desktop/src/util/config_channel_transport.c

```diff
@@ -0,0 +1,272 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <zephyr.h>
+#include <sys/byteorder.h>
+
+#include "config_channel_transport.h"
+#include "hid_report_desc.h"
+
+#define MODULE config_channel_transport
+#define TRANSPORT_HEADER_SIZE 5
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(MODULE, CONFIG_DESKTOP_CONFIG_CHANNEL_LOG_LEVEL);
+
+static const uint8_t response_pending[TRANSPORT_HEADER_SIZE] = {
+	0,		       /* Recipient */
+	0,		       /* Recipient */
+	0, 		       /* Event ID */
+	CONFIG_STATUS_PENDING, /* Status */
+	0,		       /* Event data length */
+};
+
+static int frame_length_check(const size_t length)
+{
+	const size_t min_size = TRANSPORT_HEADER_SIZE;
+	const size_t max_size = REPORT_SIZE_USER_CONFIG;
+
+	BUILD_ASSERT(min_size < max_size);
+
+	if ((length < min_size) || (length > max_size)) {
+		LOG_WRN("Unsupported report length %zu", length);
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int data_len_check(const size_t event_data_len)
+{
+	const size_t min_size = TRANSPORT_HEADER_SIZE;
+	const size_t max_size = REPORT_SIZE_USER_CONFIG;
+
+	BUILD_ASSERT(min_size < max_size);
+
+	if (event_data_len > max_size - min_size) {
+		LOG_WRN("Unsupported event data length %" PRIu8,
+			event_data_len);
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+int config_channel_report_parse(const uint8_t *buffer, size_t length,
+				struct config_event *event)
+{
+	int err = frame_length_check(length);
+
+	if (err) {
+		return err;
+	}
+
+	size_t pos = 0;
+
+	event->recipient = sys_get_le16(&buffer[pos]);
+	pos += sizeof(event->recipient);
+
+	event->event_id = buffer[pos];
+	pos += sizeof(event->event_id);
+
+	event->status = buffer[pos];
+	pos += sizeof(event->status);
+
+	uint8_t config_data_len = buffer[pos];
+	pos += sizeof(uint8_t);
+
+	__ASSERT_NO_MSG(pos == TRANSPORT_HEADER_SIZE);
+
+	err = data_len_check(config_data_len);
+	if (err) {
+		return err;
+	}
+
+	if ((length - pos) < config_data_len) {
+		LOG_ERR("Not enough data in buffer");
+		return -ENOTSUP;
+	}
+
+	__ASSERT_NO_MSG(event->dyndata.size >= config_data_len);
+	event->dyndata.size = config_data_len;
+
+	memcpy(event->dyndata.data, &buffer[pos], config_data_len);
+	pos += config_data_len;
+
+	return pos;
+}
+
+int config_channel_report_fill(uint8_t *buffer, const size_t length,
+			       const struct config_event *event)
+{
+	int err = frame_length_check(length);
+	if (!err) {
+		err = data_len_check(event->dyndata.size);
+	}
+
+	if (err) {
+		return err;
+	}
+
+	size_t pos = 0;
+
+	sys_put_le16(event->recipient, &buffer[pos]);
+	pos += sizeof(event->recipient);
+
+	buffer[pos] = event->event_id;
+	pos += sizeof(event->event_id);
+
+	buffer[pos] = event->status;
+	pos += sizeof(event->status);
+
+	buffer[pos] = (uint8_t)event->dyndata.size;
+	pos += sizeof(uint8_t); /* Data length is encoded using one byte. */
```

Comment:
`sizeof(buffer[pos])` + no comment is needed

Reply:
done

Reply:
done